### PR TITLE
Added OneVersusAll and PairwiseCoupling samples.

### PR DESCRIPTION
**Pull request by ganik**
Part of #2522.
Adds a sample for OneVersusAll classification.
Adds a sample for PairwiseCoupling classification.

### CODE REVIEW

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
+﻿using Microsoft.ML.Data;
+using Microsoft.ML.SamplesUtils;
+
+namespace Microsoft.ML.Samples.Dynamic.Trainers.MulticlassClassification
+{
+    public static class OneVersusAll
+    {
+        public static void Example()
```
Comment (Ivanidzo4ka) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs]: >Example [](start = 27, length = 7)

you probably want to link this file to extension method.

```
     /// <format type="text/markdown">
        /// <![CDATA[
        ///  [!code-csharp[SDCA](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/StochasticDualCoordinateAscentWithOptions.cs)]
        /// ]]></format>
        /// </example>
```

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            // BB       0.4919063,0.6673147,0.8326591,0.6695119,1.182151,0.230367,1.06237,1.195347,0.8771811,0.5145918
+            // CC       1.216908,1.248052,1.391902,0.4326252,1.099942,0.9262842,1.334019,1.08762,0.9468155,0.4811099
+            // DD       0.7871246,1.053327,0.8971719,1.588544,1.242697,1.362964,0.6303943,0.9810045,0.9431419,1.557455
+
+            // Create a pipeline. 
+            var pipeline =
+                    // Convert the string labels into key types.
+                    mlContext.Transforms.Conversion.MapValueToKey("Label")
+                    // Apply OneVersusAll multiclass trainer on top of SDCA Logistic Regression binary trainer.
+                    .Append(mlContext.MulticlassClassification.Trainers.OneVersusAll(mlContext.BinaryClassification.Trainers.SdcaLogisticRegression()));
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs]: usually multiclass pipelines add a MapKeyToValue at the end.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+            // Split the data into training and test sets. Only training set is used in fitting
+            // the created pipeline. Metrics are computed on the test.
+            var split = mlContext.Data.TrainTestSplit(dataView, testFraction: 0.1);
+
+            // Train the model.
+            var model = pipeline.Fit(split.TrainSet);
+
+            // Do prediction on the test set.
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs]: >Do [](start = 15, length = 2)

Generate

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+            // Evaluate the trained model using the test set.
+            var metrics = mlContext.MulticlassClassification.Evaluate(dataWithPredictions);
+            SamplesUtils.ConsoleUtils.PrintMetrics(metrics);
+
+            // Expected output:
+            //   Micro Accuracy: 0.77
+            //   Macro Accuracy: 0.75
+            //   Log Loss: 0.69
+            //   Log Loss Reduction: 0.49
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs]: >//   Log Loss Reduction: 0.49 [](start = 12, length = 29)

add reading the PredictedLabel column.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/PairwiseCoupling.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+            // Evaluate the trained model using the test set.
+            var metrics = mlContext.MulticlassClassification.Evaluate(dataWithPredictions);
+            SamplesUtils.ConsoleUtils.PrintMetrics(metrics);
+
+            // Expected output:
+            //   Micro Accuracy: 0.75
+            //   Macro Accuracy: 0.73
+            //   Log Loss: 0.70
+            //   Log Loss Reduction: 0.49
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/PairwiseCoupling.cs]: same here, add reading the PredictedLabel.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        public static void Example()
+        {
+            // Create a new context for ML.NET operations. It can be used for exception tracking and logging, 
+            // as a catalog of available operations and as the source of randomness.
+            // Setting the seed to a fixed number in this example to make outputs deterministic.
+            var mlContext = new MLContext(seed: 0);
+
+            // Create a list of data examples.
+            var examples = DatasetUtils.GenerateRandomMulticlassClassificationExamples(1000);
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs]: >GenerateRandomMulticlassClassificationExamples [](start = 40, length = 46)

is it possible to use something like GenerateRandomDataPoints that's used in binary classification?
[https://github.com/dotnet/machinelearning/blob/0ce5618516a069f173467020abcadb99a0648349/docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/FastTree.cs#L68](https://github.com/dotnet/machinelearning/blob/0ce5618516a069f173467020abcadb99a0648349/docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/FastTree.cs#L68)

**Comment by shmoradims** [comment]
> using Microsoft.ML.Data;

consider using T4 templates if you see a lot of duplicate code across multi-class samples. Below is a T4 we used for regression. All *.cs files will be autogenerated from the .tt template file.
[https://github.com/dotnet/machinelearning/pull/3099](https://github.com/dotnet/machinelearning/pull/3099)

---
Refers to: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs:1 in 01dccde. [](commit_id = 01dccdee8a4fcba4cef1f2a7a697410bda348266, deletion_comment = False)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs (hunk 6 above)
Comment (ganik) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs]: This method generates data points for multi class, with 4 labels. Except for this it does do it similarly to the GenerateRandomDataPoints. Not sure what other similarity you would want?

---
In reply to: [271461899](https://github.com/dotnet/machinelearning/pull/3159#discussion_r271461899) [](ancestors = 271461899)

**Comment by ganik** [comment]
> using Microsoft.ML.Data;

Good suggestions, done.

---
In reply to: [479163541](https://github.com/dotnet/machinelearning/pull/3159#issuecomment-479163541) [](ancestors = 479163541)

---
Refers to: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs:1 in 01dccde. [](commit_id = 01dccdee8a4fcba4cef1f2a7a697410bda348266, deletion_comment = False)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs (hunk 6 above)
Comment (ganik) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs]: this is actually done, thx

---
In reply to: [271497239](https://github.com/dotnet/machinelearning/pull/3159#discussion_r271497239) [](ancestors = 271497239,271461899)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/PairwiseCoupling.cs (hunk 5 above)
Comment (ganik) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/PairwiseCoupling.cs]: done

---
In reply to: [271136850](https://github.com/dotnet/machinelearning/pull/3159#discussion_r271136850) [](ancestors = 271136850)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs (hunk 2 above)
Comment (ganik) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs]: maybe, I dont need it here

---
In reply to: [271136620](https://github.com/dotnet/machinelearning/pull/3159#discussion_r271136620) [](ancestors = 271136620)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs (hunk 1 above)
Comment (ganik) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/OneVersusAll.cs]: no, this is about OVA not SDCA

---
In reply to: [271090986](https://github.com/dotnet/machinelearning/pull/3159#discussion_r271090986) [](ancestors = 271090986)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/NaiveBayes.cs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+                    Features = Enumerable.Repeat(label, 20).Select(x => randomFloat() + label * 0.2f).ToArray()
+                };
+            }
+        }
+
+        // Example with label and 20 feature values. A data set is a collection of such examples.
+        private class DataPoint
+        {
+            public uint Label { get; set; }
+            [VectorType(20)]
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/NaiveBayes.cs]: > [VectorType(20)] [](start = 11, length = 17)

is the annotation necessary?

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/NaiveBayes.cs [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+            // Look at 5 predictions
+            foreach (var p in predictions.Take(5))
+                Console.WriteLine($"Label: {p.Label}, Prediction: {p.PredictedLabel}");
+
+            // Evaluate the overall metrics
+            var metrics = mlContext.MulticlassClassification.Evaluate(transformedTestData);
+            SamplesUtils.ConsoleUtils.PrintMetrics(metrics);
+            
+            // Expected output:
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/NaiveBayes.cs]: >// Expected output: [](start = 12, length = 19)

how come this line is repeated?

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/MulticlassClassification.ttinclude [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+            // Run the model on test data set.
+            var transformedTestData = model.Transform(testData);
+
+            // Convert IDataView object to a list.
+            var predictions = mlContext.Data.CreateEnumerable<Prediction>(transformedTestData, reuseRowObject: false).ToList();
+
+            // Look at 5 predictions
+            foreach (var p in predictions.Take(5))
+                Console.WriteLine($"Label: {p.Label}, Prediction: {p.PredictedLabel}");
+
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/MulticlassClassification.ttinclude]: please add ExpectedOutputPerInstance after this

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/NaiveBayes.tt [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+string ClassName="NaiveBayes";
+string Trainer = "NaiveBayes";
+string MetaTrainer = null;
+string TrainerOptions = null;
+
+string OptionsInclude = "";
+string Comments= "";
+
+string ExpectedOutputPerInstance= @"// Expected output:
+            //   Label: 1, Prediction: 2
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/NaiveBayes.tt]: >Label: 1, Prediction: 2 [](start = 17, length = 23)

how come generated labels are 0,1,2 but here I see 1,2,3. how did they get changed?

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/NaiveBayes.tt [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+            //   Label: 2, Prediction: 2
+            //   Label: 3, Prediction: 2
+            //   Label: 2, Prediction: 2
+            //   Label: 3, Prediction: 2";
+
+string ExpectedOutput = @"// Expected output:
+            // Expected output:
+            // Micro Accuracy: 0.35
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/NaiveBayes.tt]: > 0.35 [](start = 30, length = 5)

can we get something above 60%

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/NaiveBayes.cs [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+            // Expected output:
+            // Micro Accuracy: 0.35
+            // Macro Accuracy: 0.33
+            // Log Loss: 34.54
+            // Log Loss Reduction: -30.47
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/NaiveBayes.cs]: we usually indent the lines below Expected output with an extra space.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/NaiveBayes.tt (hunk 11 above)
Comment (ganik) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/NaiveBayes.tt]: I tried, we cant, this is a linear model :)

---
In reply to: [271938623](https://github.com/dotnet/machinelearning/pull/3159#discussion_r271938623) [](ancestors = 271938623)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/NaiveBayes.tt (hunk 10 above)
Comment (ganik) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/NaiveBayes.tt]: no, generated labels are 1,2,3

---
In reply to: [271938136](https://github.com/dotnet/machinelearning/pull/3159#discussion_r271938136) [](ancestors = 271938136)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/NaiveBayes.cs (hunk 7 above)
Comment (ganik) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/NaiveBayes.cs]: yes, needed for schema check

---
In reply to: [271907075](https://github.com/dotnet/machinelearning/pull/3159#discussion_r271907075) [](ancestors = 271907075)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/MulticlassClassification.ttinclude [hunk 13]
```diff
@@ -0,0 +1,108 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.ML.Data;
+<# if (TrainerOptions != null) { #>
+<#=OptionsInclude#>
+<# } #>
+
+namespace Microsoft.ML.Samples.Dynamic.Trainers.MulticlassClassification
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/MulticlassClassification.ttinclude]: >Microsoft.ML [](start = 10, length = 12)

please drop Microsoft.ML prefix as per [https://github.com/dotnet/machinelearning/issues/3205](https://github.com/dotnet/machinelearning/issues/3205)

Comment (ganik) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/MulticlassClassification/MulticlassClassification.ttinclude]: done

---
In reply to: [272754353](https://github.com/dotnet/machinelearning/pull/3159#discussion_r272754353) [](ancestors = 272754353)

**Pull request merged**